sql_grader/models.py

```diff
@@ -0,0 +1,222 @@
+
+from xblock.fields import Boolean
+from xblock.fields import List
+from xblock.fields import Scope
+from xblock.fields import String
+
+
+class XBlockDataMixin(object):
+    # pylint: disable=too-few-public-methods
+
+    display_name = String(
+        display_name=_('Display Name'),
+        help=_('BLAH BLAH BE CREATIVE'),
+        default=_('SQL Problem'),
+        scope=Scope.content,
+    )
+    # TODO: Make this a choice from a list
+    dataset = String(
+        display_name=_('Dataset'),
+        help=_('Which initial dataset/database to be used for queries'),
+        default='rating',
+        scope=Scope.content,
+    )
+    answer_query = String(
+        display_name=_('Answer Query'),
+        help=_('A correct response SQL query'),
+        default='',
+        scope=Scope.content,
+        multiline_editor=True,
+    )
+    verify_query = String(
+        display_name=_('Verify Query'),
+        help=_('A verification SQL query'),
+        default='',
+        scope=Scope.content,
+        multiline_editor=True,
+    )
+    answer_result = List(
+        display_name=_('Answer Result'),
+        help=_('Results from the correct SQL query'),
+        default=None,
+        scope=Scope.content,
+    )
+    is_ordered = Boolean(
+        display_name=_('Is Ordered?'),
+        help=_('Should results be in order?'),
+        default=False,
+        scope=Scope.content,
+    )
+    editable_fields = [
+        'answer_query',
+        'dataset',
+        'display_name',
+        'verify_query',
+        'is_ordered',
+        'prompt',
+        'weight',
+    ]
+    prompt = String(
+        display_name=_('Prompt'),
+        help=_('Explanatory text to accompany the problem'),
+        default='',
+        scope=Scope.content,
+    )
+    submission_query = String(
+        display_name=_('Submission Query'),
+        help=_('A Submission Query'),
+        default='',
+        scope=Scope.user_state,
+    )
+    submission_result = List(
+        display_name=_('Submission Result'),
+        help=_('A Submission Result'),
+        default='',
+        scope=Scope.user_state,
+    )
+
+    def provide_context(self, context):  # pragma: no cover
+        """
+        Build a context dictionary to render the student view
+        """
+        context = context or {}
+        context = dict(context)
+        error_class = ''
+        if not bool(self.score) and bool(self.submission_query):
+            error_class = 'error'
+        context.update({
+            'display_name': self.display_name,
+            'prompt': self.prompt,
+            'submission_query': self.submission_query,
+            'answer': self.submission_query,
+            'score': self.score,
+            'score_weighted': int(self.score * self.weight),
+            'max_score': int(self.max_score()),
+            'user_alert': '',  # TODO: Remove this?
+            'error_class': error_class,
+        })
+        return context
+
+
+class SqlProblem(object):
```

Comment:
Note: if the use of this class seems unorthodox for an XBlock, that's because I originally wrote this in pure Python, prior to converting to an XBlock. I'm open to refactoring, if it makes sense.

Reply:
TODO: This.